packages/components/src/drop-zone/index.js

```diff
@@ -2,67 +2,190 @@
+
+			const { ownerDocument } = element;
+
+			function onDragEnter( event ) {
+				event.preventDefault();
+				if ( onDragEnterRef.current ) {
+					onDragEnterRef.current( event );
+				}
+			}
+
+			const throttledOnDragOver = throttle( ( event ) => {
+				if ( onDragOverRef.current ) {
+					onDragOverRef.current( event );
+				}
+			}, 200 );
+
+			function onDragOver( event ) {
+				// Only call onDragOver for the innermost hovered drag zones.
+				if ( ! event.defaultPrevented ) {
+					throttledOnDragOver( event );
+				}
+
+				event.preventDefault();
+			}
 
-	return {
-		...remainingState,
-		position,
-	};
+			function onDragLeave( event ) {
+				if ( onDragLeaveRef.current ) {
+					onDragLeaveRef.current( event );
+				}
+			}
+
+			function resetDragState( event ) {
+				throttledOnDragOver.cancel();
+
+				if ( isDraggingGlobally ) {
+					isDraggingGlobally = false;
+
+					element.removeAttribute( 'data-is-dragging' );
+
+					ownerDocument.addEventListener(
+						'dragover',
+						onGlobalDragOver
+					);
+
+					if ( onDragEndRef.current ) {
+						onDragEndRef.current( event );
+					}
+				}
+			}
+
+			function onDrop( event ) {
+				// This seemingly useless line has been shown to resolve a Safari issue
+				// where files dragged directly from the dock are not recognized
+				event.dataTransfer && event.dataTransfer.files.length; // eslint-disable-line no-unused-expressions
+
+				const type = getDragEventType( event );
+
+				if ( type === 'file' && onFilesDropRef.current ) {
+					onFilesDropRef.current(
+						getFilesFromDataTransfer( event.dataTransfer )
+					);
+				} else if ( type === 'html' && onHTMLDropRef.current ) {
+					onHTMLDropRef.current(
+						event.dataTransfer.getData( 'text/html' )
+					);
+				} else if ( type === 'default' && onDropRef.current ) {
+					onDropRef.current( event );
+				}
+
+				resetDragState( event );
+
+				event.stopPropagation();
+				event.preventDefault();
+			}
+
+			function onGlobalDragOver( event ) {
+				if ( ! isDraggingGlobally ) {
+					isDraggingGlobally = true;
+
+					element.setAttribute( 'data-is-dragging', 'true' );
+
+					ownerDocument.removeEventListener(
+						'dragover',
+						onGlobalDragOver
+					);
+
+					if ( onDragStartRef.current ) {
+						onDragStartRef.current( event );
+					}
+				}
+			}
+
+			element.addEventListener( 'drop', onDrop );
+			element.addEventListener( 'dragenter', onDragEnter );
+			element.addEventListener( 'dragover', onDragOver );
+			element.addEventListener( 'dragleave', onDragLeave );
+			ownerDocument.addEventListener( 'mouseup', resetDragState );
+			// Note that `dragend` doesn't fire consistently for file and HTML drag
+			// events where the drag origin is outside the browser window.
+			// In Firefox it may also not fire if the originating node is removed.
+			ownerDocument.addEventListener( 'dragend', resetDragState );
+			ownerDocument.addEventListener( 'dragover', onGlobalDragOver );
```

Comment:
Any concern about the number of event listeners this will create? With the provider it was just creating a global one right?

Reply:
No, it really won't create that much listeners. There will be listeners for each drop zone, which means each block **list** (not block!) and each media placeholder.

Reply:
We could try that, but I'd suggest trying things one step at a time. We do something similar for the in-between inserter, so I don't see why it couldn't work, though I have a feeling that we'll be much more reliant on DOM queries, which is not great. That's what I don't like about the current in-between inserter `mousemove` handlers, which I'd like to have another look at soon.

It's worth keeping in mind we used to have a drop zone for each block, which was totally crazy. :D

Reply:
> I have a feeling that we'll be much more reliant on DOM queries, which is not great.

Yeah, I think the target detection code would have to check which inner blocks rectangle it's within, so that it can isolate only child block drop targets. Having separate drop zones makes this work automatically.